Pretty print operators

xsebek opened the pull request:
- moves info about constants to `Syntax.hs`
  - adds `infoConst`&friends functions on `Const`
  - lot of noise - needs to be simplified
- changes the pretty-printing logic for `Term` application
- creates parsers for `Const` automatically
- adds a heavy dependency to get [something like `Enum`](https://hackage.haskell.org/package/finitary)

xsebek commented:
Hello @byorgey, I finally managed to get this into a somewhat working state and rebased.

It is an attempt at your third suggestion, but needs to be simplified and tested. (There are a few unit tests `>>>`.)

The no-complex-`Enum` situation is quite vexing, so I ignored it with `finitary` package that forces the issue with `Generic`. :smile: (Currently it says GPL3, but the author is fine with other licenses like Apache, in case we wanted to use it :wink: )

byorgey commented:
Great, thanks! I will take a look and provide feedback soon.

xsebek commented:
No hurry, I won't probably get back to this until the weekend. :slightly_smiling_face:

byorgey commented on src/Swarm/Language/Pretty.hs:
```diff
@@ -149,8 +156,21 @@ instance PrettyPrec Term where
+      let pL = appliedTermPrec l
+          pR = appliedTermPrec r
+      in pparens (p > pC) $ hsep
+        [ pparens (pL == pC && assoc == R) $ prettyPrec pC l
```
In the recursive calls to `prettyPrec`, you should add one to `pC` on whichever side is opposite of `assoc`.

xsebek replied:
Good idea :white_check_mark:

byorgey reviewed (commented):
Overall I like how you've organized things here.  As you say it needs a bit of cleaning up, but I think it's quite close actually.  I've left some comments on the pretty-printer which hopefully make sense.  We will definitely want to have a lot of tests for the pretty-printer in particular.  Eventually, we should definitely have some randomized testing that generates random ASTs and then pretty-prints and parses them to make sure we get back the same one we started with.  But it probably makes sense to make that into a separate issue for later (unless you really want to tackle it now).

byorgey commented on src/Swarm/Language/Syntax.hs:
```diff
@@ -167,70 +165,138 @@ data Const
+data ConstInfo = ConstInfo
+  { syntax :: Text
+  , constMeta :: ConstMeta
+  }
```
Maybe we could add `funLow` as a counterpart to `commandLow` for constructing these with one or the other value for `isCmd`.

xsebek replied:
Done :white_check_mark:

xsebek commented on src/Swarm/Language/Pretty.hs:
```diff
@@ -114,30 +114,37 @@ instance PrettyPrec Capability where
-  prettyPrec p (Arith c) = prettyPrec p c
-  prettyPrec _ Neg       = "-"
-  prettyPrec _ c         = pretty $ T.toLower (from (show c))
+  prettyPrec p c = pparens (p > precedence c) $ pretty . syntax . constInfo $ c
```
I was thinking about the `map (*)` case, that is when the operator is a parameter of some application.

xsebek replied:
OK, so *I think* that by setting the precedence of commands to 11 I get the expected output:
```haskell
>>> ppr (TApp (TVar "f") (TConst $ Say))
f say
>>> ppr (TApp (TVar "f") (TConst $ Cmp CmpEq))
f (==)
```

byorgey replied:
Ah, I see, good idea!  Actually I don't think the parser can handle this, but it should.

byorgey replied:
So this only prints operators with parents when they are on the this of an application, right? What about a bare operator by itself? Or a partially applied one?

xsebek commented on src/Swarm/Language/Syntax.hs:
```diff
@@ -167,70 +165,138 @@ data Const
+  | Arith ArithConst  -- ^ Binary arithmetic operators.
+
+  deriving stock (Eq, Ord, Generic, Data, Show)
+  deriving anyclass (Finitary)
```
Done :white_check_mark:

xsebek commented on src/Swarm/Language/Syntax.hs:
```diff
@@ -167,70 +165,138 @@ data Const
+  ConstMCmd {} -> c `notElem` [If, Force, Not, Fst, Snd]
+  _ -> False
+
+isUserCmd :: Const -> Bool
```
Renamed to `isUserFunc` :white_check_mark:

xsebek commented on src/Swarm/Language/Pretty.hs:
```diff
@@ -149,8 +156,21 @@ instance PrettyPrec Term where
+    _ -> pparens (p > 10) $
+      prettyPrec 10 t <+> prettyPrec 9 r
   prettyPrec p (TApp t1 t2)  = pparens (p > 10) $
-    prettyPrec 10 t1 <+> prettyPrec 11 t2
```
Sorry, I was probably testing what it does and forgot to revert. It should be good now :white_check_mark:

byorgey commented:
Hmm, looks like applications of unary negation don't pretty-print correctly:
```
λ> ppr (TApp (TConst Neg) (TInt 3))
(-) 3
```

byorgey commented:
This is coming along really well, I think there are just a few wrinkles to iron out.

Re: pretty-printing constants with or without parentheses, I think I understand now: by setting the precedence of commands to 11, they will never be printed with parentheses around them; binary operators will be pretty-printed specially if they are fully applied so we hit the `TApp (TApp ...)` case, but otherwise we will fall through to the bare `TConst` case in which we will just print them with parens.  I think it would probably be simpler to forget about setting the precedence of commands to 11 or whatever; instead, in the definition of `prettyPrec` for constants, just directly say `prettyPrec _ c = pparens (isBinOp c) $ ...` because whether it is an infix binary operator is really what matters.

There is still the issue of unary negation.  We need to add a special case for it first of all, so that `TApp (TConst Neg) (TInt 3)` will print correctly.  But that still leaves the issue of how to pretty-print a bare occurrence of it.  Maybe we should just adopt the same solution as Haskell, and have a synonym `negate` which is used when it is not fully applied (e.g. when it is passed as an argument to a higher-order function).

xsebek commented:
Good catch with the negation, @byorgey; I entirely forgot to handle that case.
I added unit tests for it and fixed the pretty printing:
```haskell
-3
-(-1)      -- unlike python we can not parse '--1'
-1 ^ (-2)  -- the inner repr is -(...) and secondly '...^-2' would not parse
-(1 + -2)  -- this parses fine in swarm (haskell dislikes this)
```

xsebek commented:
About `negate` - I think that just can be a nice function in user-space:
```haskell
> (\i. -i)
\i. -i : int -> int
```
We don't currently support passing operator sections as parameters, so it does not seem like a problem.

byorgey commented:
@xsebek  , congrats on having your first PR accepted to Swarm, and thanks for all the other contributions (discussion, bug reports, etc.) so far as well! We're really glad to welcome you as part of the [community](https://github.com/byorgey/swarm/blob/main/COMMUNITY.md).  In a minute we will send you an invite to the repo (see https://github.com/byorgey/swarm/blob/main/CONTRIBUTING.md#i-have-push-access-to-the-swarm-repository-now-what ).  If you use IRC, feel free to also join the `#swarm` channel on Libera.Chat.
